[Refactor] Make params of `UI#showText` sane

DayKev opened the pull request:
## What are the changes the user will see?
N/A

## Why am I making these changes?
The parameters of `showText` were cursed.

## What are the changes from a developer perspective?
The parameters for `UI#showText` and all similar functions were changed to use an object for the optional parameters.

The `delay` and `callback` parameters of `UI#showDialogue` were swapped so that `delay` could be made optional.

## How to test the changes?
Do things that generate messages I guess?

## Checklist
- [x] Otherwise: **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes manually?
- [x] Are all unit tests still passing? (`pnpm test:silent`)
  - [x] Have I created new automated tests (`pnpm test:create`) or updated existing tests related to the PR's changes?

DayKev committed:
Update `showText`

DayKev committed:
Update `showDialogue`

DayKev commented:
It'll probably be easier to review by hiding whitespace-only changes (such as via https://github.com/Despair-Games/poketernity/pull/1272/files?diff=unified&w=1).

DayKev committed:
Apply Biome

innerthunder commented on src/ui/handlers/message-ui-handler.ts:
```diff
@@ -74,35 +68,27 @@ export abstract class MessageUiHandler extends AwaitableUiHandler {
-    callbackDelay?: number | null,
-    prompt?: boolean | null,
-    promptDelay?: number | null,
+    { delay = 20, callback, callbackDelay, prompt, promptDelay }: ShowTextOptions = {},
```
```suggestion
    { delay = 20, callback, callbackDelay = 0, prompt = false, promptDelay = 0 }: ShowTextOptions = {},
```
These defaults are already in place with the truthy/falsy checks in here; we might as well make them more explicit.

innerthunder commented on src/data/mystery-encounters/encounters/safari-zone-encounter.ts:
```diff
@@ -204,9 +204,15 @@ const safariZoneGameOptions: MysteryEncounterOption[] = [
       const fleeChangeResult = tryChangeFleeStage(1, 8);
       if (!fleeChangeResult) {
-        await showEncounterText(getEncounterText(`${namespace}:safari.busy_eating`) ?? "", null, 1000, false);
+        await showEncounterText(getEncounterText(`${namespace}:safari.busy_eating`) ?? "", {
+          callbackDelay: 1000,
+          prompt: false,
+        });
```
`prompt` is set to the default here (and a few other places)

DayKev replied:
`prompt` defaults to `true` for `showEncounterText`, so this is different from the default.

innerthunder commented on src/data/mystery-encounters/utils/encounter-dialogue-utils.ts:
```diff
@@ -57,16 +57,23 @@ export function queueEncounterMessage(contentKey: string): void {
+    callbackDelay?: number;
+    prompt?: boolean;
+    promptDelay?: number;
+  } = {},
```
You could use `Omit<ShowTextOptions, "callback">` as the argument type instead.

DayKev replied:
Created an interface right above this instead so I can put different TSDocs on it, since the default for `prompt` is different from `showText`.

innerthunder commented on src/data/mystery-encounters/utils/encounter-pokemon-utils.ts:
```diff
@@ -790,10 +783,7 @@ export async function catchPokemon(
-        doPokemonCatchMenu,
-        0,
-        true,
+        { callback: doPokemonCatchMenu, callbackDelay: 0, prompt: true },
```
Here `callBackDelay` is set to the default value.

DayKev committed:
Add explicit defaults to `MessageUiHandler#showTextInternal`

- Change `if (numberVar)` to `if (numberVar > 0)`

- Remove passing redundant default `callbackDelay` value

- Create `ShowEncounterTextOptions` in `encounter-dialogue-utils.ts`

- Move some TSDocs from `MessageUiHandler#showText`
to `ShowTextOptions`

MokaStitcher commented on src/@types/ui-types.ts:
```diff
@@ -5,6 +5,31 @@ import type { PlayerPokemon } from "#field/player-pokemon";
 
+export interface ShowTextOptions {
+  /**
+   * The delay in milliseconds before the dialogue is displayed.
```
I think we need to be careful about using dialogue VS text here. Dialogue in the codebase specifically refers to trainer/story dialogue, which have a speaker name and get automatically cut based on $ signs and can include emotes information and other things, other text does not necessarily do that. Showtext is used for any text like "Charmander used Ember", tutorials, etc.
I think it should be text here?

DayKev replied:
I just copy/pasted the existing TSDocs, moving them to the interface from the function.

MokaStitcher commented on src/@types/ui-types.ts:
```diff
@@ -5,6 +5,31 @@ import type { PlayerPokemon } from "#field/player-pokemon";
+   * @defaultValue `20`
+   */
+  delay?: number;
+  /** A callback function to execute after the dialogue is displayed. */
```
```suggestion
  /** A callback function to execute after the text is done displaying. */
```
is that correct? Is it called when the text has finished appearing one letter at a time?

DayKev replied:
I assume so.

MokaStitcher commented on src/data/mystery-encounters/utils/encounter-dialogue-utils.ts:
```diff
@@ -49,24 +49,41 @@ export function queueEncounterMessage(contentKey: string): void {
 
+interface ShowEncounterTextOptions {
+  /**
+   * The delay in milliseconds before the dialogue is displayed.
```
```suggestion
   * The delay in milliseconds before the text starts getting displayed.
```

MokaStitcher commented on src/data/mystery-encounters/utils/encounter-dialogue-utils.ts:
```diff
@@ -49,24 +49,41 @@ export function queueEncounterMessage(contentKey: string): void {
+   */
+  delay?: number;
+  /**
+   * The delay in milliseconds before executing the callback.
```
which callback are we talking about?

DayKev replied:
I guess it means the `() => resolve()` being passed to `showText()`? This is also just moving existing TSDocs, I didn't write anything new/different to what already existed.
I suspect these docs were also originally just copy/pasted from the `showText()` docs.

MokaStitcher replied:
Oh, I thought those didn't have doc :plains

DayKev replied:
The docs were only on the one in `MessageUiHandler` lol.

MokaStitcher reviewed (commented):
<sub>I made a few comments earlier, don't remember what it was about but I can't continue reviewing now, so sending what I have

DayKev committed:
Fix merge issue

DayKev committed:
Update TSDocs

MokaStitcher commented on src/data/mystery-encounters/encounters/clowning-around-encounter.ts:
```diff
@@ -418,13 +418,13 @@ export const ClowningAroundEncounter: MysteryEncounter = MysteryEncounterBuilder
 
     await globalScene.ui.setMessageMode();
-    await showEncounterText(`${namespace}:option.1.ability_prompt`, null, 500, false);
+    await showEncounterText(`${namespace}:option.1.ability_prompt`, { callbackDelay: 500, prompt: false });
```
why is there a callbackDelay since there's no callback and no prompt 😭
Based on the MessageUiHandler.showText code we probably shouldn't touch it, but I think we need to look deeper into what those delays and things actually do, they all seem entangled with each other. 
Probably not for this PR I don't know

DayKev replied:
Should I add a todo comment?

DayKev replied:
Actually all `showEncounterText` calls have a callback, it passes `() => resolve()` to `ui.showText()`.

MokaStitcher commented on src/data/mystery-encounters/encounters/delibirdy-encounter.ts:
```diff
@@ -153,12 +153,7 @@ export const DelibirdyEncounter: MysteryEncounter = MysteryEncounterBuilder.with
-            undefined,
-            true,
-          );
+          await showEncounterText(i18next.t("battle:rewardGain", { modifierName: shellBell.name }), { prompt: true });
```
since you said prompt is true by default for showEncounterText in another comment
```suggestion
          await showEncounterText(i18next.t("battle:rewardGain", { modifierName: shellBell.name }));
```

MokaStitcher commented on src/data/mystery-encounters/encounters/global-trade-system-encounter.ts:
```diff
@@ -505,7 +505,7 @@ async function doTradeOptionPhaseCallback(): Promise<void> {
   await showTradeBackground();
   await doPokemonTradeSequence(tradedPokemon, newPlayerPokemon);
-  await showEncounterText(`${namespace}:trade_received`, null, 0, true, 4000);
+  await showEncounterText(`${namespace}:trade_received`, { callbackDelay: 0, prompt: true, promptDelay: 4000 });
```
```suggestion
  await showEncounterText(`${namespace}:trade_received`, { promptDelay: 4000 });
```

MokaStitcher commented on src/data/mystery-encounters/encounters/global-trade-system-encounter.ts:
```diff
@@ -195,7 +195,7 @@ export const GlobalTradeSystemEncounter: MysteryEncounter = MysteryEncounterBuil
                   + getNatureName(tradePokemon.getNature())
                   + (formName ? "     |     " + i18next.t("pokemonInfoContainer:form") + " " + formName : "");
-                showEncounterText(`${line1}\n${line2}`, 0, 0, false);
+                showEncounterText(`${line1}\n${line2}`, { delay: 0, callbackDelay: 0, prompt: false });
```
```suggestion
                showEncounterText(`${line1}\n${line2}`, { delay: 0, prompt: false });
```

MokaStitcher commented on src/data/mystery-encounters/encounters/delibirdy-encounter.ts:
```diff
@@ -334,12 +319,7 @@ export const DelibirdyEncounter: MysteryEncounter = MysteryEncounterBuilder.with
-            undefined,
-            true,
-          );
+          await showEncounterText(i18next.t("battle:rewardGain", { modifierName: shellBell.name }), { prompt: true });
```
```suggestion
          await showEncounterText(i18next.t("battle:rewardGain", { modifierName: shellBell.name }));
```

MokaStitcher commented on src/data/mystery-encounters/encounters/delibirdy-encounter.ts:
```diff
@@ -252,12 +242,7 @@ export const DelibirdyEncounter: MysteryEncounter = MysteryEncounterBuilder.with
-              undefined,
-              true,
-            );
+            await showEncounterText(i18next.t("battle:rewardGain", { modifierName: shellBell.name }), { prompt: true });
```
```suggestion
            await showEncounterText(i18next.t("battle:rewardGain", { modifierName: shellBell.name }));
```

MokaStitcher commented on src/data/mystery-encounters/encounters/delibirdy-encounter.ts:
```diff
@@ -234,12 +229,7 @@ export const DelibirdyEncounter: MysteryEncounter = MysteryEncounterBuilder.with
-              undefined,
-              true,
-            );
+            await showEncounterText(i18next.t("battle:rewardGain", { modifierName: shellBell.name }), { prompt: true });
```
```suggestion
            await showEncounterText(i18next.t("battle:rewardGain", { modifierName: shellBell.name }));
```

MokaStitcher commented on src/data/mystery-encounters/encounters/safari-zone-encounter.ts:
```diff
@@ -205,9 +205,15 @@ const safariZoneGameOptions: MysteryEncounterOption[] = [
+        await showEncounterText(getEncounterText(`${namespace}:safari.busy_eating`) ?? "", {
+          callbackDelay: 1000,
+          prompt: false,
+        });
```
```suggestion
        const text = getEncounterText(`${namespace}:safari.busy_eating`) ?? "";
        await showEncounterText(text, { callbackDelay: 1000, prompt: false });
```
easier to read imo

MokaStitcher commented on src/data/mystery-encounters/encounters/safari-zone-encounter.ts:
```diff
@@ -232,11 +238,11 @@ const safariZoneGameOptions: MysteryEncounterOption[] = [
+        getEncounterText(catchChangeResult ? `${namespace}:safari.angry` : `${namespace}:safari.beside_itself_angry`)
+          ?? "",
+        { callbackDelay: 1000, prompt: false },
+      );
```
```suggestion
      const textKey = catchChangeResult ? `${namespace}:safari.angry` : `${namespace}:safari.beside_itself_angry`;
      await showEncounterText(getEncounterText(textKey) ?? "", { callbackDelay: 1000, prompt: false });
```

MokaStitcher commented on src/data/mystery-encounters/encounters/the-strong-stuff-encounter.ts:
```diff
@@ -178,7 +178,7 @@ export const TheStrongStuffEncounter: MysteryEncounter = MysteryEncounterBuilder
       encounter.setDialogueToken("reductionValue", HIGH_BST_REDUCTION_VALUE.toString());
       encounter.setDialogueToken("increaseValue", BST_INCREASE_VALUE.toString());
-      await showEncounterText(`${namespace}:option.1.selected_2`, null, undefined, true);
+      await showEncounterText(`${namespace}:option.1.selected_2`, { prompt: true });
```
```suggestion
      await showEncounterText(`${namespace}:option.1.selected_2`);
```

MokaStitcher commented on src/data/mystery-encounters/encounters/safari-zone-encounter.ts:
```diff
@@ -517,6 +523,12 @@ function isPokemonFlee(pokemon: EnemyPokemon, fleeStage: number): boolean {
 }
 
+/**
+ * TODO: confirm function description (it probably changes how likely the pokemon is to flee?)
```
changing the fleeing chance is probably the point I'd say. Safari catch rates are different because they use a safari ball and bait/stones mechanics (at least in mainline)
Still worth looking into so it can be documented fully though

MokaStitcher commented on src/data/mystery-encounters/encounters/training-session-encounter.ts:
```diff
@@ -278,7 +278,7 @@ export const TrainingSessionEncounter: MysteryEncounter = MysteryEncounterBuilde
                 },
                 onHover: () => {
-                  showEncounterText(ability.description, 0, 0, false);
+                  showEncounterText(ability.description, { delay: 0, callbackDelay: 0, prompt: false });
```
```suggestion
                  showEncounterText(ability.description, { delay: 0, prompt: false });
```

MokaStitcher commented on src/data/mystery-encounters/encounters/trash-to-treasure-encounter.ts:
```diff
@@ -125,12 +125,9 @@ export const TrashToTreasureEncounter: MysteryEncounter = MysteryEncounterBuilde
-          );
+          await showEncounterText(i18next.t("battle:rewardGain", { modifierName: modifier.type.name }), {
+            prompt: true,
+          });
```
```suggestion
          await showEncounterText(i18next.t("battle:rewardGain", { modifierName: modifier.type.name }));
```

MokaStitcher commented on src/data/mystery-encounters/encounters/trash-to-treasure-encounter.ts:
```diff
@@ -216,12 +213,9 @@ async function tryApplyDigRewardItems() {
-  );
+  await showEncounterText(i18next.t("battle:rewardGainCount", { modifierName: leftovers.name, count: 2 }), {
+    prompt: true,
+  });
```
```suggestion
  await showEncounterText(i18next.t("battle:rewardGainCount", { modifierName: leftovers.name, count: 2 }));
```

MokaStitcher commented on src/data/mystery-encounters/encounters/trash-to-treasure-encounter.ts:
```diff
@@ -252,12 +246,9 @@ async function tryApplyDigRewardItems() {
-  );
+  await showEncounterText(i18next.t("battle:rewardGainCount", { modifierName: shellBell.name, count: 2 }), {
+    prompt: true,
+  });
```
```suggestion
  await showEncounterText(i18next.t("battle:rewardGainCount", { modifierName: shellBell.name, count: 2 }));
```

MokaStitcher commented on src/data/mystery-encounters/encounters/weird-dream-encounter.ts:
```diff
@@ -283,7 +283,7 @@ export const WeirdDreamEncounter: MysteryEncounter = MysteryEncounterBuilder.wit
       );
 
-      await showEncounterText(`${namespace}:option.2.selected_2`, null, undefined, true);
+      await showEncounterText(`${namespace}:option.2.selected_2`, { prompt: true });
```
```suggestion
      await showEncounterText(`${namespace}:option.2.selected_2`);
```

MokaStitcher commented on src/data/mystery-encounters/utils/encounter-phase-utils.ts:
```diff
@@ -606,7 +606,11 @@ export function selectPokemonForOption(
-                      showEncounterText(i18next.t("mysteryEncounterMessages:cancel_option"), 0, 0, false);
+                      showEncounterText(i18next.t("mysteryEncounterMessages:cancel_option"), {
+                        delay: 0,
+                        callbackDelay: 0,
```
```suggestion
```

MokaStitcher commented on src/data/mystery-encounters/utils/encounter-phase-utils.ts:
```diff
@@ -747,7 +751,11 @@ export function selectOptionThenPokemon(
-          showEncounterText(i18next.t("mysteryEncounterMessages:cancel_option"), 0, 0, false);
+          showEncounterText(i18next.t("mysteryEncounterMessages:cancel_option"), {
+            delay: 0,
+            callbackDelay: 0,
```
```suggestion
```

MokaStitcher commented on src/data/mystery-encounters/utils/encounter-pokemon-utils.ts:
```diff
@@ -770,13 +767,9 @@ export async function catchPokemon(
             },
           };
           const promptRelease = () => {
```
do we really need a buffer function with a single line in it lol
```suggestion
```

DayKev replied:
It's used in multiple places above so I can't just delete it.
For some reason it's defined *after* it's used a couple times, which is... lol.

MokaStitcher commented on src/phases/encounter-phase.ts:
```diff
@@ -627,13 +633,12 @@ export class EncounterPhase extends BattlePhase {
     handleTutorial(Tutorial.ACCESS_MENU).then(() => super.end());
   }
 
+  // TODO: something different
```
🥲

MokaStitcher commented on src/phases/encounter-phase.ts:
```diff
@@ -653,22 +658,22 @@ export class EncounterPhase extends BattlePhase {
             gameData.saveSeenDialogue(localizationKey);
           }
-          ui.showDialogue(encounterDialogue, enemy?.species.name, null, () => {
+          ui.showDialogue(encounterDialogue, enemy?.species.name ?? "something broke", () => {
```
```suggestion
          ui.showDialogue(encounterDialogue, enemy?.species.name ?? "MissingNo", () => {
```
😏

MokaStitcher commented on src/phases/encounter-phase.ts:
```diff
@@ -653,22 +658,22 @@ export class EncounterPhase extends BattlePhase {
   }
 
-  /**
+  /*
```
why not keep it as tsdoc?

DayKev replied:
That's weird, how did this get changed.

DayKev replied:
Oh wait I see why I did that. Okay, lemme change this.

DayKev replied:
Okay, added TSDocs to the methods and left this as a non-TSDoc.

MokaStitcher commented on src/phases/end-card-phase.ts:
```diff
@@ -45,16 +45,13 @@ export class EndCardPhase extends Phase {
-        null,
-        true,
-      );
+        prompt: true,
```
we actually probably don't want the prompt here, and want a delay of 0, since this is used to clear the text box
<sub> this is done in so many spots in the ui :'(

DayKev replied:
Should I leave it as-is and put a todo comment, or remove the `prompt` param?

DayKev replied:
Added the todo comment for now.

MokaStitcher commented on src/phases/faint-phase.ts:
```diff
@@ -268,10 +268,9 @@ export class FaintPhase extends PokemonPhase {
   protected handleFinalBossFaint(): void {
     const enemy = this.getPokemon();
+    const text = classicFinalBossDialogue.secondStageWin;
     if (enemy.formIndex > 0) {
```
```suggestion
    if (enemy.formIndex > 0) {
      const text = classicFinalBossDialogue.secondStageWin;
```

MokaStitcher commented on src/ui/handlers/message-ui-handler.ts:
```diff
@@ -74,35 +64,27 @@ export abstract class MessageUiHandler extends AwaitableUiHandler {
-    prompt?: boolean | null,
-    promptDelay?: number | null,
+    delay?: number,
+    callback?: VoidFunction,
```
since the order of these 2 parameters were flipped in the UI class, should we flip it here as well to reduce potential confusion?

DayKev replied:
Ugh is the order not consistent between different version of this function... yeah they should all match.

DayKev replied:
Huh, `callback` isn't optional in `UI#showDialogue` but it is here...

MokaStitcher reviewed (commented):
mostly removing some params with default values in ME code
<sub>this was exhausting to go through

DayKev committed:
Remove redundant params

DayKev committed:
Update docs in `encounter-phase.ts`

DayKev committed:
Match order of params between all versions of `showDialogue()`

DayKev committed:
Simplify code in `safari-zone-encounter.ts`

DayKev committed:
Remove `!` in `encounter-phase.ts`

DayKev committed:
Merge branch 'beta' into refactor/showText-params

DayKev committed:
Add `TODO` comment to `end-card-phase.ts
